src/jax_finufft/shapes.py

```diff
@@ -116,10 +115,15 @@ def abstract_eval(source, *points, output_shape, **_):
     assert all(p.ndim == 2 for p in points)
     assert all(p.shape == points[0].shape for p in points[1:])
     assert source.shape[0] == points[0].shape[0]
-    if output_shape is None:
+
+    if output_shape is None:  # Type 2
         assert source.ndim == 2 + ndim
-        return ShapedArray(source.shape[:2] + (points[0].shape[-1],), source_dtype)
-    else:
+        return source.update(
+            shape=source.shape[:2] + (points[0].shape[-1],), dtype=source_dtype
+        )
```

Comment:
This is an unrelated change that I noticed. We're no longer supposed to manually construct a `ShapedArray` here because it should return the same type as the input (tracer vs concrete value).